File: SparkScalaCourse/src/main/scala/com/leo/spark/FriendsByAge.scala
package com.leo.scalaspark

import org.apache.log4j._
import org.apache.spark._

object FriendsByAge {

  /** Split and return a tuple with age and number of friends for every line */
  def parseLine(line: String): (Int, Int) ={
    // split every line by the comma
    val linesColumns = line.split(",")
    // get the third field of the line (age) and transform to int
    val age = linesColumns(2).toInt
    // get the fourth field of the line (number of friends) and transform to int
    val numFriends = linesColumns(3).toInt
    // return a tuple (age, number of friends)
    (age, numFriends)
  }

  def main(args: Array[String]): Unit ={

    val DATA_PATH = "../data/"

    // Remove messages that are not error
    Logger.getLogger("org").setLevel(Level.ERROR)

    // The function
    val friendsByAgeFinal = new SparkContext("local[*]", "FriendsByAge")
      .textFile(DATA_PATH + "fakefriends-noheader.csv")
      .map(parseLine)
      .mapValues(x=>(x,1))
      .reduceByKey((x, y) => (x._1 + y._1, x._2 + y._2))
      .mapValues(x=>(x._1 / x._2))
      .sortByKey()

    friendsByAgeFinal.collect().foreach(println)

    // Creates new spark context FriendsByAge locally
    //val sc = new SparkContext("local[*]", "FriendsByAge")

    // Read the data file
    // val dataFile = sc.textFile(DATA_PATH + "fakefriends-noheader.csv")

    //xx.collect().foreach(f=>{
    //  println(f)})

    // Get every line of the file as a input (with map) to parseLine
    //val relevantData = dataFile.map(parseLine)

    // Create a tuple (age, (number of friends, 1)) for every sample
    //val tupleData = relevantData.mapValues(x=>(x,1))

    // For a tuple (key, value) group the same keys and sum its values
    //val groupedByAgeData = tupleData.reduceByKey((x, y) => (x._1 + y._1, x._2 + y._2))

    // Divide the total amount to number of occurrences

    //val averageByAge = groupedByAgeData.mapValues(x=>(x._1 / x._2))

    // Sort by key

    //val results = averageByAge.collect()
    //results.sorted.foreach(println)
  }
}

File: SparkPythonCourse/friends-by-age.py
# -*- coding: utf-8 -*-
"""
Created on Fri Dec 24 12:15:12 2021

@author: LeonardoDiasdaRosa
"""

from pyspark import SparkConf, SparkContext

## Data
DATA_PATH = '../data/'
FILE = "fakefriends-noheader.csv"

data = DATA_PATH + FILE

# Function to get data from lines
def parseLine(lines):
    line = lines.split(",")
    age = int(line[2])
    friends = int(line[3])
    return (age, friends)


## Create Spark context
conf = SparkConf().setMaster("local").setAppName("FriendsByAge")
sc = SparkContext(conf = conf)

## Friends by age
friendsByAge = sc.textFile(data).map(parseLine).mapValues(lambda x: (x, 1))\
    .reduceByKey(lambda x, y: (x[0] + y[0], x[1] + y[1])).mapValues(lambda x: x[0] / x[1])

for i in (sorted(friendsByAge.collect())):
    print(i[0], round(i[1], 2))

## Read file
#lines = sc.textFile(data)

## Data
#dataValues = lines.map(parseLine)

## Adjust dict
#newDict = dataValues.mapValues(lambda x: (x, 1))

## Sum the occurences
#sumDict = newDict.reduceByKey(lambda x, y: (x[0] + y[0], x[1] + y[1]))
#averagesByAge = sumDict.mapValues(lambda x: x[0] / x[1])

## Print
#for i in (sorted(averagesByAge.collect())):
#    print(i[0], round(i[1], 2))
